### DISCUSSION TITLE
Toptica iBeam smart integration with microscope

### DISCUSSION THREAD

**Discussion by jacopoabramo** [ImSwitch/ImSwitch#105] [General]
Greetings,

I successfully integrated the Toptica iBeam smart laser into ImSwitch by using the microscope layer. The laser manager works fine, but there is an issue: as reported in this [issue](https://github.com/python-microscope/microscope/issues/233), due to a change in the distribution package `pyserial` in version 3.5, I had to downgrade my local version to 3.4 to make it work. I wanted to push the changes and eventually ask for a merge request but I see that from the `setup.cfg` file in ImSwitch that the minimum version required during installation is 3.5. Do you think it would be possible to merge the changes by downgrading the `pyserial` version to 3.4 or are there any other device which require 3.5 as a minimum version (as I'm suspecting)?

Thanks!

**Comment by xcasas** [comment]
Hi @jacopoabramo,

We use pyserial for lantz-dev and I made a temporary branch to test pyserial 3.4 and it seems to pass the tests (except some tests that fail sometimes and I'm trying to investigate why, but independently of pyserial), and I quickly tried in my setup with some lasers and it seems to work fine. So go ahead :)

How did you do it btw, do you have the repo somewhere? Did you do a LaserManager for the Toptica iBeam or generalized to any python-microscope supported laser?

**Reply by jacopoabramo** [reply]
Hey @kasasxav , apologies I forgot to push the changes into a new branch. You can find the integration at the following:

https://github.com/Eggeling-Lab-Microscope-Software/ImSwitch/blob/toptica/imswitch/imcontrol/model/managers/lasers/TopticaLaserManager.py

It's a simple laser manager class, but I can follow up on your suggestion and try to create a more general `PyMicroscopeLaserManager` class (I also have a Cobolt laser in my setup which I can use to try and control it). I don't think I'll be able to make it so that it can also support remote devices, which I think it would be cool to integrate, but maybe you have more insight on this?

**Reply by xcasas** [reply]
Cool!! Looks super nice. 
I don't have much experience with remote devices. But I'm thinking what if in managerProperties there could be a field called "driver" or whatever.
Then in the import you can `import microscope.lights`, then in the json file `driver:"topica.TopicaiBeam"`,
so then in line 18 you can do something like `eval(["microscope.lights."+ managerProperties["driver"] + "(" str(managerProperties["digitalPorts"]) ")"])` 

I think there are better ways than eval, but something like that would be super cool and would work with other python microscope light sources. What do you think?

**Reply by jacopoabramo** [reply]
Alright, I think I was able to find a nice way to create a generic class to make use of python-microscope.

In [here ](https://github.com/Eggeling-Lab-Microscope-Software/ImSwitch/blob/toptica/imswitch/imcontrol/model/managers/lasers/PyMicroscopeLaserManager.py), I used `__import__` and `getattr` to create an instance of the TopticaiBeam laser, and it seems to be working. I've tried with the Cobolt laser too, but there seems to be something wrong with microscope when trying to instantiate it - it's out of the scope of ImSwitch though. Any thoughts?

I'm thinking that for remote devices a separate manager should be created, it's probably best to differentiate them in order to not overcomplicate this class.

**Reply by xcasas** [reply]
Looks great! much better with `__import__` and `getattr`. What error does it give you? If you want you can do a PR and I can check it also with the cobolts we have.

I agree with the remote devices!

**Reply by jacopoabramo** [reply]
I just opened the following [issue](https://github.com/python-microscope/microscope/issues/234), but yeah I'll perform a PR. I didn't write any unit tests for this though, I hope it won't be a problem.

**Reply by xcasas** [reply]
I think it's fine for now :) if there are no lasers does python microscope start a mock? In that case just adding an example file would work because we have a unit test that checks that the example setups are ran correctly, but no worries go ahead.

**Reply by xcasas** [reply]
And yep seems to be a PM issue
